### PR TITLE
Pyomo.DoE bug fixes

### PR DESCRIPTION
**Pull request by adowling2**
## Summary/Motivation:
* Minor enhancements in Pyomo.DoE in preparation for a tutorial/workshop

## Changes proposed in this PR:
- Fixed error about ambiguous `a.any()` versus `a.all()` for logical check of a list-like object
- Added a TypeError to prevent a downstream cryptic error message
- Update type checking per suggested best practices from @jsiirola and @mrmundt 
- Improved initialization of the Jacobian and Cholesky factorization
- Added optionally passing `theta` parameter values into `create_model` when generating scenarios. This is especially helpful if `create_model` includes initialization such was using the simulator in `pyomo.dae`
- Added more comments
- Fixed degree of freedom issue by fixing the parameters on the nominal case model. This is important if `create_model` creates a full model in `stage1` model. Our original design anticipated `stage1` mode would be really lite, but that is inconvenient when trying to use the same `create_model` for `parmest` and `pyomo.doe`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/doe/doe.py [hunk 1]
```diff
@@ -101,6 +101,8 @@ def __init__(
         """
 
         # parameters
+        if type(param_init) != dict:
```
Comment (mrmundt) [inline pyomo/contrib/doe/doe.py:104]: Maybe a better way to do this is to use:
```
if not isinstance(param_init, dict):
```

File: pyomo/contrib/doe/measurements.py [hunk 2]
```diff
@@ -177,20 +177,20 @@ def _check_valid_input(
...
-        if values and len(values) != len_indices:
+        if values is not None and len(values) != len_indices:
             raise ValueError("Values is of different length with indices.")
 
         if (
-            lower_bounds
-            and type(lower_bounds) == list
-            and len(lower_bounds) != len_indices
+            lower_bounds is not None # ensure not None
+            and type(lower_bounds) == list # ensure list
```
Comment (mrmundt) [inline pyomo/contrib/doe/measurements.py:185]: Similar to above: ``isinstance`` might be better

File: pyomo/contrib/doe/measurements.py [hunk 3]
```diff
@@ -177,20 +177,20 @@ def _check_valid_input(
...
+            and len(lower_bounds) != len_indices # ensure same length
         ):
             raise ValueError("Lowerbounds is of different length with indices.")
 
         if (
-            upper_bounds
-            and type(upper_bounds) == list
-            and len(upper_bounds) != len_indices
+            upper_bounds is not None # ensure None
+            and type(upper_bounds) == list # ensure list
```
Comment (mrmundt) [inline pyomo/contrib/doe/measurements.py:192]: Similar to above: ``isinstance`` might be better

**Review by mrmundt — DISMISSED** [review]

File: pyomo/contrib/doe/doe.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/contrib/doe/doe.py:104]: In other parts of Pyomo, we use `Config*` objects to help with option validation, but for these cases, I think `isinstance` is appropriate.

That said, unless you really, really need this to be a `dict` subclass, I would recommend `isinstance(param_init, collections.Mapping)` (which allows for any mapping type - even if they don't inherit from `dict`).  Similarly, I would test against `collections.Sequence` and not `list` in `measurements.py` (which admits other sequences like `tuple`)

**Comment by adowling2** [comment]
@jsiirola @mrmundt Suggestions for `type(lower_bounds) in [int, float]`?

**Comment by adowling2** [comment]
Any reasons for not replacing `if type(self.L_initial) != type(None)` with `if self.L_initial is not None`?

**Comment by jsiirola** [comment]
For checking for numeric types, we generally use
```python
from pyomo.common.numeric_types import native_numeric_types

if type(lower_bound) in native_numeric_types:
   pass
# or
if lower_bound.__class__ in native_numeric_types:
   pass
```

and yes: `if self.L_initial is not None` is the recommended idiom.

**Comment by mrmundt** [comment]
@adowling2 - Please run `black -S -C` on your changes to enforce PEP8 standards.

**Comment by adowling2** [comment]
More development plans:
- [ ] Ensure `discretize_model = None` works on TCLab example (separate repo, really helpful for testing)
- [ ] Add automated initialization (assumes user initialized model in `create_model`)
  - [ ] Add toggle for feature
  - [ ] Initialize sensitivity calculation
  - [ ] Initialize FIM calculation
  - [ ] Initialize Cholesky factorization

**Comment by adowling2** [comment]
> @adowling2 - Please run `black -S -C` on your changes to enforce PEP8 standards.

I got `Error: no such option: -C`

However, `black -S` worked fine. I have `version 19.10b0` if that matters.

**Comment by adowling2** [comment]
Update: I reinstalled black and it works using the command provided by @mrmundt. Now using `black, 24.4.2`. I'm going to keep the comment above for my future reference.

**Comment by adowling2** [comment]
@dlakes94 @slilonfe5 @jialuw96 Check out these enhancements to Pyomo.DoE. We'll want to use this branch (which will hopefully be merged into main soon) when refreshing the Pyomo.DoE examples.

**Comment by adowling2** [comment]
Anything else needed to get this reviewed and merged?

File: pyomo/contrib/doe/doe.py [hunk 4]
```diff
@@ -1011,6 +1070,32 @@ def fim_rule(m, p, q):
...
+            # Assemble the FIM matrix
+            fim = np.zeros((len(self.param), len(self.param)))
+            for i, bu in enumerate(m.regression_parameters):
+                for j, un in enumerate(m.regression_parameters):
+                    fim[i][j] = m.fim[bu, un].value
+
+            # Calculate the eigenvalues of the FIM matrix
+            eig = np.linalg.eigvals(fim)
+
+            # If the smallest eigenvalue is (pratcially) negative, add a diagonal matrix to make it positive definite
```
Comment (blnicho) [inline pyomo/contrib/doe/doe.py:1086]:
```suggestion
            # If the smallest eigenvalue is (practically) negative, add a diagonal matrix to make it positive definite
```

**Review by blnicho — APPROVED** [review]
I found one small typo but otherwise this looks fine.

**Review by mrmundt — APPROVED** [review]

**Pull request merged**